### ISSUE TITLE
Kubernetes service discovery intermittently not populating the pod IP into `__address__`

### ISSUE THREAD

**Issue by zoonage** [prometheus/prometheus#11600]
### What did you do?

Notes before reading:

1. I've not been able to replicate this issue reliably as I think it might be a race condition of some kind.
2. In the example below the pod was removed before Prometheus had time to fix the address, however we do see an eventual reconciliation. If needed I can track down more examples. We see this happen >20 times a day on some of our Prometheus servers.

In the graph below we can see that this pod has the address `:9090` for the `prometheus_build_info`. This `trigger-campaigns` pod does not publish this metric and is in fact from the Prometheus server scraping itself.

<img width="1424" alt="image" src="https://user-images.githubusercontent.com/7557073/202756352-757f71b2-e0e8-4336-837d-0e65789771a0.png">

In the graph below we can see that for the whole lifetime of this pod it's IP is `10.130.73.162`

<img width="1414" alt="image" src="https://user-images.githubusercontent.com/7557073/202756420-4779637b-994a-4a6e-b2f5-1dae44c566d8.png">

I've done some digging into this already and my current suspicion is something along the lines of

1. Prometheus receives an update from the endpoints watch
2. Prometheus receives an incomplete, or otherwise not quite correct update from the pods watch
3. `pe.pod.Status.PodIP` from [this](https://github.com/prometheus/prometheus/blob/main/discovery/kubernetes/endpoints.go#L276) line (indicating an issue with the endpoint reflector?) or [this](https://github.com/prometheus/prometheus/blob/main/discovery/kubernetes/endpoints.go#L378) line (indicating an issue with the pod reflector? Because we have the `container` label on this metric I'm assuming it must be reaching this stage.) is an empty string, causing the `:9090`.

I've done some experimenting with using the k8s discovery directly but haven't been able to replicate the issue. I've included the code I've used for this below.

```
package main

import (
	"context"
	"os"

	"github.com/go-kit/log"
	"github.com/prometheus/prometheus/discovery"
	"github.com/prometheus/prometheus/discovery/kubernetes"
	"github.com/prometheus/prometheus/discovery/targetgroup"
	"go.uber.org/zap"
)

func main() {
	il, _ := zap.NewProduction()
	logger := il.Sugar()
	logger.Info("Testing")
	sdConfig := kubernetes.SDConfig{
		KubeConfig: os.Getenv("KUBECONFIG"),
		Role:       "endpoints",
	}
	d, err := sdConfig.NewDiscoverer(discovery.DiscovererOptions{
		Logger: log.NewJSONLogger(os.Stdout),
	})
	if err != nil {
		logger.Fatalw("could not create discoverer", "error", err)
	}
	tchan := make(chan []*targetgroup.Group)
	go func() {
		d.Run(context.Background(), tchan)
	}()
	for {
		targets := <-tchan
		for _, targetGroup := range targets {
			targetAddresses := make([]string, 0)
			for _, target := range targetGroup.Targets {
				address := target["__address__"]
				if address[0:1] == ":" {
					targetAddresses = append(targetAddresses, string(address))
				}
			}
			if len(targetAddresses) != 0 {
				logger.Infow("Found targets", "group", targetGroup.Source, "adresses", targetAddresses)
			}
		}
	}
}
```

We only see this issue on 4 of our 8 Prometheus servers running in production.

Unforunately we cannot replicate this behaviour as reliably in our test environments (one occurrence in the last 2 weeks) so I believe it is in some way related to load.

My first thought is to add a `if !pe.pod.Status.PodIP { continue }` statement and see if that still has the issue to confirm my suspicion from earlier.

Eventually the `address` does get corrected to the right value if the pod is in the service discovery for long enough. I suspect it is when the [sync interval](https://github.com/prometheus/prometheus/blob/main/discovery/kubernetes/kubernetes.go#L385) comes around however I've not been able to confirm this as I don't know how to see when this sync is triggered.

### What did you expect to see?

The `address` label as `10.130.73.162:9090` and metrics from the pod being scraped by Prometheus.

### What did you see instead? Under which circumstances?

The `address` label as `:9090` and metrics from the Prometheus server doing the scrape being scraped.

### System information

Amazon Linux 2 5.4.219-126.411.amzn2.x86_64 on AWS EKS 1.21

### Prometheus version

```text
Build Information
version	2.34.0
revision	881111fec4332c33094a6fb2680c71fffc427275
branch	HEAD
buildUser	root@121ad7ea5487
buildDate	20220315-15:18:00
goVersion	go1.17.8
```

### Prometheus configuration file

```yaml
Generated by Prometheus operator


- job_name: serviceMonitor/prod/trigger-campaigns-welcome-processor-low-priority/0
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [job]
    separator: ;
    regex: (.*)
    target_label: __tmp_prometheus_job_name
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_label_app_kubernetes_io_instance, __meta_kubernetes_service_labelpresent_app_kubernetes_io_instance]
    separator: ;
    regex: (trigger-campaigns-welcome-processor-low-priority);true
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_label_app_kubernetes_io_name, __meta_kubernetes_service_labelpresent_app_kubernetes_io_name]
    separator: ;
    regex: (trigger-campaigns);true
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    separator: ;
    regex: "9090"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Node;(.*)
    target_label: node
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Pod;(.*)
    target_label: pod
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: service
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_container_name]
    separator: ;
    regex: (.*)
    target_label: container
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: job
    replacement: ${1}
    action: replace
  - separator: ;
    regex: (.*)
    target_label: endpoint
    replacement: "9090"
    action: replace
  - source_labels: [__meta_kubernetes_pod_label_ometria_com_tribe]
    separator: ;
    regex: (.*)
    target_label: tribe
    replacement: $1
    action: replace
  - separator: ;
    regex: (.*)
    target_label: rule_namespace
    replacement: prod
    action: replace
  - source_labels: [__address__]
    separator: ;
    regex: (.*)
    modulus: 1
    target_label: __tmp_hash
    replacement: $1
    action: hashmod
  - source_labels: [__tmp_hash]
    separator: ;
    regex: "0"
    replacement: $1
    action: keep
  metric_relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: rule_namespace
    replacement: prod
    action: replace
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
    namespaces:
      own_namespace: false
      names:
      - prod
```
```


### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

**Comment by zoonage** [comment]
Some following steps we're going to try on our side once we are out of code freeze (start of December)

1. Upgrade Prometheus to the latest version, we're running the latest in our test and internal tooling environment at the moment to see if we see this happening, although as mentioned in the ticket there are significantly fewer instances in those clusters anyway
2. Run a modified version of Prometheus with extra logging or a reduced refresh interval to force sync SD configs faster

**Comment by machine424** [comment]
Should we close this as fixed by https://github.com/prometheus/prometheus/pull/11642?

**Comment by zoonage** [comment]
> Should we close this as fixed by #11642?

I don't work at the company that was affected by this bug anymore so I'm unable to validate it directly

**Comment by machine424** [comment]
Thanks for your response.
I think the tests you added are sufficient to validate this.
I think we can close this as fixed by https://github.com/prometheus/prometheus/pull/11642 @roidelapluie
